Patch 57 -- export audit-round3 GLES texture readback from the clean NO-ISLAND chain.
--- a/dlls/wined3d/texture_gl.c	2026-08-18 00:23:12.512647215 +0300
+++ b/dlls/wined3d/texture_gl.c
@@ -1872,7 +1872,123 @@
     }
 }
 
-static void wined3d_texture_gl_download_data_slow_path(struct wined3d_texture_gl *texture_gl,
+static BOOL mgs_gles_texture_download_data(struct wined3d_texture_gl *texture_gl,
+        unsigned int sub_resource_idx, struct wined3d_context_gl *context_gl,
+        const struct wined3d_bo_address *data, const struct wined3d_format_gl *format_gl)
+{
+    struct wined3d_bo_gl *bo = wined3d_bo_gl(data->buffer_object);
+    const struct wined3d_gl_info *gl_info = context_gl->gl_info;
+    unsigned int level = sub_resource_idx % texture_gl->t.level_count;
+    unsigned int width = wined3d_texture_get_level_width(&texture_gl->t, level);
+    unsigned int height = wined3d_texture_get_level_height(&texture_gl->t, level);
+    unsigned int row_pitch, ignored_slice_pitch, y, x;
+    GLenum target = wined3d_texture_gl_get_sub_resource_target(texture_gl, sub_resource_idx);
+    BOOL bgr = format_gl->f.id == WINED3DFMT_B8G8R8A8_UNORM
+            || format_gl->f.id == WINED3DFMT_B8G8R8X8_UNORM;
+    BOOL rgba = format_gl->f.id == WINED3DFMT_R8G8B8A8_UNORM;
+    GLint old_pack_alignment, old_pack_row_length, old_pack_skip_rows;
+    GLint old_pack_skip_pixels, old_pack_buffer;
+    GLuint old_read_fbo = context_gl->fbo_read_binding;
+    GLuint old_draw_fbo = context_gl->fbo_draw_binding;
+    struct fbo_entry *old_current_fbo = context_gl->current_fbo;
+    BOOL old_rebind_fbo = context_gl->rebind_fbo;
+    GLint old_read_buffer;
+    GLenum status;
+    BYTE *pixels;
+
+    if (target != GL_TEXTURE_2D || !gl_info->is_gles
+            || (texture_gl->t.resource.format_attrs & WINED3D_FORMAT_ATTR_COMPRESSED)
+            || format_gl->f.download || format_gl->f.byte_count != 4 || (!bgr && !rgba)
+            || !(pixels = malloc(width * height * 4)))
+        return FALSE;
+
+    wined3d_texture_get_pitch(&texture_gl->t, level, &row_pitch, &ignored_slice_pitch);
+    gl_info->gl_ops.gl.p_glGetIntegerv(GL_READ_BUFFER, &old_read_buffer);
+    wined3d_context_gl_apply_fbo_state_explicit(context_gl, GL_READ_FRAMEBUFFER,
+            &texture_gl->t.resource, sub_resource_idx, NULL, 0, WINED3D_LOCATION_TEXTURE_RGB);
+    MGS2_GL_INFO(gl_info)->gl_ops.gl.p_glReadBuffer(GL_COLOR_ATTACHMENT0);
+    status = gl_info->fbo_ops.glCheckFramebufferStatus(GL_READ_FRAMEBUFFER);
+    if (status != GL_FRAMEBUFFER_COMPLETE)
+    {
+        ERR("GLES texture readback FBO is incomplete: %#x.\n", status);
+        gl_info->fbo_ops.glBindFramebuffer(GL_READ_FRAMEBUFFER, old_read_fbo);
+        gl_info->fbo_ops.glBindFramebuffer(GL_DRAW_FRAMEBUFFER, old_draw_fbo);
+        context_gl->fbo_read_binding = old_read_fbo;
+        context_gl->fbo_draw_binding = old_draw_fbo;
+        context_gl->current_fbo = old_current_fbo;
+        context_gl->rebind_fbo = old_rebind_fbo;
+        MGS2_GL_INFO(gl_info)->gl_ops.gl.p_glReadBuffer(old_read_buffer);
+        free(pixels);
+        return FALSE;
+    }
+    gl_info->gl_ops.gl.p_glGetIntegerv(GL_PACK_ALIGNMENT, &old_pack_alignment);
+    gl_info->gl_ops.gl.p_glGetIntegerv(GL_PACK_ROW_LENGTH, &old_pack_row_length);
+    gl_info->gl_ops.gl.p_glGetIntegerv(GL_PACK_SKIP_ROWS, &old_pack_skip_rows);
+    gl_info->gl_ops.gl.p_glGetIntegerv(GL_PACK_SKIP_PIXELS, &old_pack_skip_pixels);
+    gl_info->gl_ops.gl.p_glGetIntegerv(GL_PIXEL_PACK_BUFFER_BINDING, &old_pack_buffer);
+    MGS2_GL_INFO(gl_info)->gl_ops.gl.p_glPixelStorei(GL_PACK_ALIGNMENT, 1);
+    MGS2_GL_INFO(gl_info)->gl_ops.gl.p_glPixelStorei(GL_PACK_ROW_LENGTH, 0);
+    MGS2_GL_INFO(gl_info)->gl_ops.gl.p_glPixelStorei(GL_PACK_SKIP_ROWS, 0);
+    MGS2_GL_INFO(gl_info)->gl_ops.gl.p_glPixelStorei(GL_PACK_SKIP_PIXELS, 0);
+    GL_EXTCALL(glBindBuffer(GL_PIXEL_PACK_BUFFER, 0));
+    MGS2_GL_INFO(gl_info)->gl_ops.gl.p_glReadPixels(0, 0, width, height,
+            GL_RGBA, GL_UNSIGNED_BYTE, pixels);
+
+    if (bgr)
+    {
+        for (y = 0; y < height; ++y)
+        {
+            BYTE *pixel = pixels + y * width * 4;
+            for (x = 0; x < width; ++x, pixel += 4)
+            {
+                BYTE tmp = pixel[0];
+                pixel[0] = pixel[2];
+                pixel[2] = tmp;
+            }
+        }
+    }
+
+    if (bo)
+    {
+        GL_EXTCALL(glBindBuffer(GL_PIXEL_PACK_BUFFER, bo->id));
+        for (y = 0; y < height; ++y)
+            GL_EXTCALL(glBufferSubData(GL_PIXEL_PACK_BUFFER,
+                    (GLintptr)data->addr + bo->b.buffer_offset + y * row_pitch,
+                    width * 4, pixels + y * width * 4));
+        wined3d_context_gl_reference_bo(context_gl, bo);
+    }
+    else
+    {
+        for (y = 0; y < height; ++y)
+        {
+            BYTE *dst = (BYTE *)data->addr + y * row_pitch;
+            const BYTE *src = pixels + y * width * 4;
+            for (x = 0; x < width; ++x)
+            {
+                memcpy(dst, src, 4);
+                dst += 4; src += 4;
+            }
+        }
+    }
+
+    MGS2_GL_INFO(gl_info)->gl_ops.gl.p_glPixelStorei(GL_PACK_ALIGNMENT, old_pack_alignment);
+    MGS2_GL_INFO(gl_info)->gl_ops.gl.p_glPixelStorei(GL_PACK_ROW_LENGTH, old_pack_row_length);
+    MGS2_GL_INFO(gl_info)->gl_ops.gl.p_glPixelStorei(GL_PACK_SKIP_ROWS, old_pack_skip_rows);
+    MGS2_GL_INFO(gl_info)->gl_ops.gl.p_glPixelStorei(GL_PACK_SKIP_PIXELS, old_pack_skip_pixels);
+    GL_EXTCALL(glBindBuffer(GL_PIXEL_PACK_BUFFER, old_pack_buffer));
+    gl_info->fbo_ops.glBindFramebuffer(GL_READ_FRAMEBUFFER, old_read_fbo);
+    gl_info->fbo_ops.glBindFramebuffer(GL_DRAW_FRAMEBUFFER, old_draw_fbo);
+    context_gl->fbo_read_binding = old_read_fbo;
+    context_gl->fbo_draw_binding = old_draw_fbo;
+    context_gl->current_fbo = old_current_fbo;
+    context_gl->rebind_fbo = old_rebind_fbo;
+    MGS2_GL_INFO(gl_info)->gl_ops.gl.p_glReadBuffer(old_read_buffer);
+    checkGLcall("GLES texture readback");
+    free(pixels);
+    return TRUE;
+}
+
+static BOOL wined3d_texture_gl_download_data_slow_path(struct wined3d_texture_gl *texture_gl,
         unsigned int sub_resource_idx, struct wined3d_context_gl *context_gl, const struct wined3d_bo_address *data)
 {
     struct wined3d_bo_gl *bo = wined3d_bo_gl(data->buffer_object);
@@ -1888,13 +2004,18 @@
 
     format_gl = wined3d_format_gl(texture_gl->t.resource.format);
 
+    if (mgs_gles_texture_download_data(texture_gl, sub_resource_idx, context_gl, data, format_gl))
+        return TRUE;
+    if (gl_info->is_gles)
+        return FALSE;
+
     /* Only support read back of converted P8 textures. */
     if (texture_gl->t.flags & WINED3D_TEXTURE_CONVERTED && format_gl->f.id != WINED3DFMT_P8_UINT
             && !format_gl->f.download)
     {
         ERR("Trying to read back converted texture %p, %u with format %s.\n",
                 texture_gl, sub_resource_idx, debug_d3dformat(format_gl->f.id));
-        return;
+        return FALSE;
     }
 
     sub_resource = &texture_gl->t.sub_resources[sub_resource_idx];
@@ -1906,7 +2027,7 @@
         if (format_gl->f.download)
         {
             FIXME("Reading back converted array texture %p is not supported.\n", texture_gl);
-            return;
+            return FALSE;
         }
 
         WARN_(d3d_perf)("Downloading all miplevel layers to get the data for a single sub-resource.\n");
@@ -1914,7 +2035,7 @@
         if (!(temporary_mem = calloc(texture_gl->t.layer_count, sub_resource->size)))
         {
             ERR("Out of memory.\n");
-            return;
+            return FALSE;
         }
     }
 
@@ -1939,7 +2060,7 @@
         if (!(temporary_mem = malloc(src_slice_pitch)))
         {
             ERR("Failed to allocate memory.\n");
-            return;
+            return FALSE;
         }
     }
 
@@ -1975,15 +2096,15 @@
         TRACE("Downloading texture %p, %u, level %u, format %#x, type %#x, data %p.\n",
                 texture_gl, sub_resource_idx, level, format_gl->format, format_gl->type, mem);
 
-        /* glGetTexImage() does not exist in GLES and is NULL here; the only
-         * readback path is an FBO plus glReadPixels(), which this function is
-         * not shaped for. Skip rather than jump to zero -- the destination
-         * keeps whatever it already held, which is wrong but survivable,
-         * whereas the call is fatal. */
         if (gl_info->is_gles)
-            FIXME("Texture download is not implemented on GLES.\n");
+        {
+            ERR("Unsupported GLES converted texture readback format %s.\n",
+                    debug_d3dformat(format_gl->f.id));
+            free(temporary_mem);
+            return FALSE;
+        }
         else
-            gl_info->gl_ops.gl.p_glGetTexImage(target, level, format_gl->format, format_gl->type, mem);
+            MGS2_GL_INFO(gl_info)->gl_ops.gl.p_glGetTexImage(target, level, format_gl->format, format_gl->type, mem);
         checkGLcall("glGetTexImage");
     }
 
@@ -2019,9 +2140,10 @@
     }
 
     free(temporary_mem);
+    return TRUE;
 }
 
-static void wined3d_texture_gl_download_data(struct wined3d_context *context,
+static BOOL wined3d_texture_gl_download_data_internal(struct wined3d_context *context,
         struct wined3d_texture *src_texture, unsigned int src_sub_resource_idx, unsigned int src_location,
         const struct wined3d_box *src_box, const struct wined3d_bo_address *dst_bo_addr,
         const struct wined3d_format *dst_format, unsigned int dst_x, unsigned int dst_y, unsigned int dst_z,
@@ -2051,7 +2173,7 @@
     else if (src_location != WINED3D_LOCATION_TEXTURE_RGB)
     {
         FIXME("Unhandled location %s.\n", wined3d_debug_location(src_location));
-        return;
+        return FALSE;
     }
 
     src_level = src_sub_resource_idx % src_texture->level_count;
@@ -2062,13 +2184,13 @@
             || src_box->front || src_box->back != src_depth)
     {
         FIXME("Unhandled source box %s.\n", debug_box(src_box));
-        return;
+        return FALSE;
     }
 
     if (dst_x || dst_y || dst_z)
     {
         FIXME("Unhandled destination (%u, %u, %u).\n", dst_x, dst_y, dst_z);
-        return;
+        return FALSE;
     }
 
     if (dst_format->id != src_texture->resource.format->id)
@@ -2076,7 +2198,7 @@
         FIXME("Unhandled format conversion (%s -> %s).\n",
                 debug_d3dformat(src_texture->resource.format->id),
                 debug_d3dformat(dst_format->id));
-        return;
+        return FALSE;
     }
 
     wined3d_texture_get_pitch(src_texture, src_level, &src_row_pitch, &src_slice_pitch);
@@ -2084,7 +2206,7 @@
     {
         FIXME("Unhandled destination pitches %u/%u (source pitches %u/%u).\n",
                 dst_row_pitch, dst_slice_pitch, src_row_pitch, src_slice_pitch);
-        return;
+        return FALSE;
     }
 
     wined3d_texture_gl_bind_and_dirtify(src_texture_gl, context_gl, srgb);
@@ -2097,8 +2219,8 @@
             || (src_texture->flags & WINED3D_TEXTURE_CONVERTED)))
             || target == GL_TEXTURE_1D_ARRAY)
     {
-        wined3d_texture_gl_download_data_slow_path(src_texture_gl, src_sub_resource_idx, context_gl, dst_bo_addr);
-        return;
+        return wined3d_texture_gl_download_data_slow_path(src_texture_gl, src_sub_resource_idx,
+                context_gl, dst_bo_addr);
     }
 
     if (format_gl->f.conv_byte_count)
@@ -2106,7 +2228,7 @@
         FIXME("Attempting to download a converted texture, type %s format %s.\n",
                 debug_d3dresourcetype(src_texture->resource.type),
                 debug_d3dformat(format_gl->f.id));
-        return;
+        return FALSE;
     }
 
     if ((dst_bo = dst_bo_addr->buffer_object))
@@ -2135,9 +2257,16 @@
                 src_texture, src_sub_resource_idx, src_level, format_gl->format, format_gl->type, offset);
 
         if (gl_info->is_gles)
-            FIXME("Texture download is not implemented on GLES.\n");
+        {
+            if (!mgs_gles_texture_download_data(src_texture_gl, src_sub_resource_idx,
+                    context_gl, dst_bo_addr, format_gl))
+            {
+                ERR("Unsupported GLES texture readback format %s.\n", debug_d3dformat(format_gl->f.id));
+                return FALSE;
+            }
+        }
         else
-            gl_info->gl_ops.gl.p_glGetTexImage(target, src_level, format_gl->format, format_gl->type, offset);
+            MGS2_GL_INFO(gl_info)->gl_ops.gl.p_glGetTexImage(target, src_level, format_gl->format, format_gl->type, offset);
         checkGLcall("glGetTexImage");
     }
 
@@ -2147,1048 +2276,20 @@
         wined3d_context_gl_reference_bo(context_gl, wined3d_bo_gl(dst_bo));
         checkGLcall("glBindBuffer");
     }
-}
-
-/* Context activation is done by the caller. */
-static BOOL wined3d_texture_gl_load_sysmem(struct wined3d_texture_gl *texture_gl,
-        unsigned int sub_resource_idx, struct wined3d_context_gl *context_gl, uint32_t dst_location)
-{
-    struct wined3d_texture_sub_resource *sub_resource;
-
-    sub_resource = &texture_gl->t.sub_resources[sub_resource_idx];
-
-    /* We cannot download data from multisample textures directly. */
-    if (wined3d_texture_gl_is_multisample_location(texture_gl, WINED3D_LOCATION_TEXTURE_RGB)
-            || sub_resource->locations & WINED3D_LOCATION_RB_MULTISAMPLE)
-        wined3d_texture_load_location(&texture_gl->t, sub_resource_idx, &context_gl->c, WINED3D_LOCATION_RB_RESOLVED);
-
-    if (sub_resource->locations & WINED3D_LOCATION_RB_RESOLVED)
-    {
-        texture2d_read_from_framebuffer(&texture_gl->t, sub_resource_idx, &context_gl->c,
-                WINED3D_LOCATION_RB_RESOLVED, dst_location);
-        return TRUE;
-    }
-
-    if (sub_resource->locations & (WINED3D_LOCATION_TEXTURE_RGB | WINED3D_LOCATION_TEXTURE_SRGB))
-    {
-        unsigned int row_pitch, slice_pitch, level;
-        struct wined3d_bo_address data;
-        struct wined3d_box src_box;
-        uint32_t src_location;
-
-        level = sub_resource_idx % texture_gl->t.level_count;
-        wined3d_texture_get_bo_address(&texture_gl->t, sub_resource_idx, &data, dst_location);
-        src_location = sub_resource->locations & WINED3D_LOCATION_TEXTURE_RGB
-                ? WINED3D_LOCATION_TEXTURE_RGB : WINED3D_LOCATION_TEXTURE_SRGB;
-        wined3d_texture_get_level_box(&texture_gl->t, level, &src_box);
-        wined3d_texture_get_pitch(&texture_gl->t, level, &row_pitch, &slice_pitch);
-        wined3d_texture_gl_download_data(&context_gl->c, &texture_gl->t, sub_resource_idx, src_location,
-                &src_box, &data, texture_gl->t.resource.format, 0, 0, 0, row_pitch, slice_pitch);
-
-        ++texture_gl->t.download_count;
-        return TRUE;
-    }
-
-    if (sub_resource->locations & WINED3D_LOCATION_DRAWABLE)
-    {
-        texture2d_read_from_framebuffer(&texture_gl->t, sub_resource_idx, &context_gl->c,
-                texture_gl->t.resource.draw_binding, dst_location);
-        return TRUE;
-    }
-
-    FIXME("Can't load texture %p, %u with location flags %s into sysmem.\n",
-            texture_gl, sub_resource_idx, wined3d_debug_location(sub_resource->locations));
-
-    return FALSE;
-}
-
-static BOOL wined3d_texture_load_drawable(struct wined3d_texture *texture,
-        unsigned int sub_resource_idx, struct wined3d_context *context)
-{
-    struct wined3d_device *device;
-    unsigned int level;
-    RECT r;
-
-    if (texture->resource.bind_flags & WINED3D_BIND_DEPTH_STENCIL)
-    {
-        uint32_t current = texture->sub_resources[sub_resource_idx].locations;
-        FIXME("Unimplemented copy from %s for depth/stencil buffers.\n",
-                wined3d_debug_location(current));
-        return FALSE;
-    }
-
-    if (wined3d_resource_is_offscreen(&texture->resource))
-    {
-        ERR("Trying to load offscreen texture into WINED3D_LOCATION_DRAWABLE.\n");
-        return FALSE;
-    }
-
-    device = texture->resource.device;
-    level = sub_resource_idx % texture->level_count;
-    SetRect(&r, 0, 0, wined3d_texture_get_level_width(texture, level),
-            wined3d_texture_get_level_height(texture, level));
-    wined3d_texture_load_location(texture, sub_resource_idx, context, WINED3D_LOCATION_TEXTURE_RGB);
-    device->blitter->ops->blitter_blit(device->blitter, WINED3D_BLIT_OP_COLOR_BLIT, context,
-            texture, sub_resource_idx, WINED3D_LOCATION_TEXTURE_RGB, &r,
-            texture, sub_resource_idx, WINED3D_LOCATION_DRAWABLE, &r,
-            NULL, WINED3D_TEXF_POINT, NULL);
 
     return TRUE;
 }
 
-static BOOL wined3d_texture_load_renderbuffer(struct wined3d_texture *texture,
-        unsigned int sub_resource_idx, struct wined3d_context *context, uint32_t dst_location)
-{
-    unsigned int level = sub_resource_idx % texture->level_count;
-    const RECT rect = {0, 0,
-            wined3d_texture_get_level_width(texture, level),
-            wined3d_texture_get_level_height(texture, level)};
-    struct wined3d_texture_sub_resource *sub_resource;
-    uint32_t src_location, locations;
-
-    sub_resource = &texture->sub_resources[sub_resource_idx];
-    locations = sub_resource->locations;
-    if (texture->resource.bind_flags & WINED3D_BIND_DEPTH_STENCIL)
-    {
-        FIXME("Unimplemented copy from %s for depth/stencil buffers.\n",
-                wined3d_debug_location(locations));
-        return FALSE;
-    }
-
-    if (locations & WINED3D_LOCATION_RB_MULTISAMPLE)
-        src_location = WINED3D_LOCATION_RB_MULTISAMPLE;
-    else if (locations & WINED3D_LOCATION_RB_RESOLVED)
-        src_location = WINED3D_LOCATION_RB_RESOLVED;
-    else if (locations & WINED3D_LOCATION_TEXTURE_SRGB)
-        src_location = WINED3D_LOCATION_TEXTURE_SRGB;
-    else if (locations & WINED3D_LOCATION_TEXTURE_RGB)
-        src_location = WINED3D_LOCATION_TEXTURE_RGB;
-    else if (locations & WINED3D_LOCATION_DRAWABLE)
-        src_location = WINED3D_LOCATION_DRAWABLE;
-    else /* texture2d_blt_fbo() will load the source location if necessary. */
-        src_location = WINED3D_LOCATION_TEXTURE_RGB;
-
-    texture2d_blt_fbo(texture->resource.device, context, WINED3D_TEXF_POINT, texture,
-            sub_resource_idx, src_location, &rect, texture, sub_resource_idx, dst_location, &rect, NULL);
-
-    return TRUE;
-}
-
-static BOOL wined3d_texture_gl_load_texture(struct wined3d_texture_gl *texture_gl,
-        unsigned int sub_resource_idx, struct wined3d_context_gl *context_gl, bool srgb)
-{
-    struct wined3d_device *device = texture_gl->t.resource.device;
-    const struct wined3d_gl_info *gl_info = context_gl->gl_info;
-    unsigned int level, src_row_pitch, src_slice_pitch;
-    struct wined3d_texture_sub_resource *sub_resource;
-    const struct wined3d_format *format;
-    struct wined3d_bo_address data;
-    struct wined3d_box src_box;
-    uint32_t dst_location;
-    bool depth;
-
-    depth = texture_gl->t.resource.bind_flags & WINED3D_BIND_DEPTH_STENCIL;
-    sub_resource = &texture_gl->t.sub_resources[sub_resource_idx];
-    level = sub_resource_idx % texture_gl->t.level_count;
-    wined3d_texture_get_level_box(&texture_gl->t, level, &src_box);
-
-    if (!depth && sub_resource->locations & (WINED3D_LOCATION_TEXTURE_SRGB | WINED3D_LOCATION_TEXTURE_RGB)
-            && (texture_gl->t.resource.format_caps & WINED3D_FORMAT_CAP_FBO_ATTACHABLE_SRGB)
-            && fbo_blitter_supported(WINED3D_BLIT_OP_COLOR_BLIT, gl_info,
-                    &texture_gl->t.resource, WINED3D_LOCATION_TEXTURE_RGB,
-                    &texture_gl->t.resource, WINED3D_LOCATION_TEXTURE_SRGB))
-    {
-        RECT src_rect;
-
-        SetRect(&src_rect, src_box.left, src_box.top, src_box.right, src_box.bottom);
-        if (srgb)
-            texture2d_blt_fbo(device, &context_gl->c, WINED3D_TEXF_POINT,
-                    &texture_gl->t, sub_resource_idx, WINED3D_LOCATION_TEXTURE_RGB, &src_rect,
-                    &texture_gl->t, sub_resource_idx, WINED3D_LOCATION_TEXTURE_SRGB, &src_rect, NULL);
-        else
-            texture2d_blt_fbo(device, &context_gl->c, WINED3D_TEXF_POINT,
-                    &texture_gl->t, sub_resource_idx, WINED3D_LOCATION_TEXTURE_SRGB, &src_rect,
-                    &texture_gl->t, sub_resource_idx, WINED3D_LOCATION_TEXTURE_RGB, &src_rect, NULL);
-
-        return TRUE;
-    }
-
-    if (!depth && sub_resource->locations & (WINED3D_LOCATION_RB_MULTISAMPLE | WINED3D_LOCATION_RB_RESOLVED)
-            && (!srgb || (texture_gl->t.resource.format_caps & WINED3D_FORMAT_CAP_FBO_ATTACHABLE_SRGB)))
-    {
-        uint32_t src_location = sub_resource->locations & WINED3D_LOCATION_RB_RESOLVED ?
-                WINED3D_LOCATION_RB_RESOLVED : WINED3D_LOCATION_RB_MULTISAMPLE;
-        RECT src_rect;
-
-        SetRect(&src_rect, src_box.left, src_box.top, src_box.right, src_box.bottom);
-        dst_location = srgb ? WINED3D_LOCATION_TEXTURE_SRGB : WINED3D_LOCATION_TEXTURE_RGB;
-        if (fbo_blitter_supported(WINED3D_BLIT_OP_COLOR_BLIT, gl_info,
-                &texture_gl->t.resource, src_location, &texture_gl->t.resource, dst_location))
-            texture2d_blt_fbo(device, &context_gl->c, WINED3D_TEXF_POINT, &texture_gl->t, sub_resource_idx,
-                    src_location, &src_rect, &texture_gl->t, sub_resource_idx, dst_location, &src_rect, NULL);
-
-        return TRUE;
-    }
-
-    /* Upload from system memory */
-
-    if (srgb)
-    {
-        dst_location = WINED3D_LOCATION_TEXTURE_SRGB;
-        if ((sub_resource->locations & (WINED3D_LOCATION_TEXTURE_RGB | texture_gl->t.resource.map_binding))
-                == WINED3D_LOCATION_TEXTURE_RGB)
-        {
-            FIXME_(d3d_perf)("Downloading RGB texture %p, %u to reload it as sRGB.\n", texture_gl, sub_resource_idx);
-            wined3d_texture_load_location(&texture_gl->t, sub_resource_idx,
-                    &context_gl->c, texture_gl->t.resource.map_binding);
-        }
-    }
-    else
-    {
-        dst_location = WINED3D_LOCATION_TEXTURE_RGB;
-        if ((sub_resource->locations & (WINED3D_LOCATION_TEXTURE_SRGB | texture_gl->t.resource.map_binding))
-                == WINED3D_LOCATION_TEXTURE_SRGB)
-        {
-            FIXME_(d3d_perf)("Downloading sRGB texture %p, %u to reload it as RGB.\n", texture_gl, sub_resource_idx);
-            wined3d_texture_load_location(&texture_gl->t, sub_resource_idx,
-                    &context_gl->c, texture_gl->t.resource.map_binding);
-        }
-    }
-
-    if (!(sub_resource->locations & wined3d_texture_sysmem_locations))
-    {
-        WARN("Trying to load a texture from sysmem, but no simple location is valid.\n");
-        /* Lets hope we get it from somewhere... */
-        wined3d_texture_load_location(&texture_gl->t, sub_resource_idx, &context_gl->c, WINED3D_LOCATION_SYSMEM);
-    }
-
-    wined3d_texture_get_pitch(&texture_gl->t, level, &src_row_pitch, &src_slice_pitch);
-
-    format = texture_gl->t.resource.format;
-
-    wined3d_texture_get_memory(&texture_gl->t, sub_resource_idx, &context_gl->c, &data);
-    wined3d_texture_gl_upload_data(&context_gl->c, wined3d_const_bo_address(&data), format, &src_box,
-            src_row_pitch, src_slice_pitch, &texture_gl->t, sub_resource_idx, dst_location, 0, 0, 0);
-    return TRUE;
-}
-
-static void wined3d_texture_gl_prepare_buffer_object(struct wined3d_texture_gl *texture_gl,
-        unsigned int sub_resource_idx, struct wined3d_context_gl *context_gl)
-{
-    struct wined3d_texture_sub_resource *sub_resource;
-    struct wined3d_bo_gl *bo;
-
-    sub_resource = &texture_gl->t.sub_resources[sub_resource_idx];
-
-    if (sub_resource->bo)
-        return;
-
-    if (!(bo = malloc(sizeof(*bo))))
-        return;
-
-    if (!wined3d_device_gl_create_bo(wined3d_device_gl(texture_gl->t.resource.device),
-            context_gl, sub_resource->size, GL_PIXEL_UNPACK_BUFFER, GL_STREAM_DRAW, true,
-            GL_MAP_READ_BIT | GL_MAP_WRITE_BIT | GL_CLIENT_STORAGE_BIT, bo))
-    {
-        free(bo);
-        return;
-    }
-
-    TRACE("Created buffer object %u for texture %p, sub-resource %u.\n", bo->id, texture_gl, sub_resource_idx);
-    sub_resource->bo = &bo->b;
-}
-
-static bool wined3d_texture_use_immutable_storage(const struct wined3d_texture *texture,
-        const struct wined3d_gl_info *gl_info)
-{
-    /* We don't expect to create texture views for textures with height-scaled formats.
-     * Besides, ARB_texture_storage doesn't allow specifying exact sizes for all levels. */
-    return gl_info->supported[ARB_TEXTURE_STORAGE]
-            && !(texture->resource.format_attrs & WINED3D_FORMAT_ATTR_HEIGHT_SCALE);
-}
-
-void wined3d_texture_gl_prepare_texture(struct wined3d_texture_gl *texture_gl,
-        struct wined3d_context_gl *context_gl, bool srgb)
-{
-    uint32_t alloc_flag = srgb ? WINED3D_TEXTURE_SRGB_ALLOCATED : WINED3D_TEXTURE_RGB_ALLOCATED;
-    const struct wined3d_gl_info *gl_info = context_gl->gl_info;
-    struct wined3d_resource *resource = &texture_gl->t.resource;
-    const struct wined3d_format *format = resource->format;
-    const struct wined3d_format_gl *format_gl;
-    GLenum internal;
-
-    TRACE("texture_gl %p, context_gl %p, srgb %d, format %s.\n",
-            texture_gl, context_gl, srgb, debug_d3dformat(format->id));
-
-    if (texture_gl->t.flags & alloc_flag)
-        return;
-
-    if (resource->format_caps & WINED3D_FORMAT_CAP_DECOMPRESS)
-    {
-        TRACE("WINED3D_FORMAT_CAP_DECOMPRESS set.\n");
-        texture_gl->t.flags |= WINED3D_TEXTURE_CONVERTED;
-        format = wined3d_resource_get_decompress_format(resource);
-    }
-    else if (format->conv_byte_count)
-    {
-        texture_gl->t.flags |= WINED3D_TEXTURE_CONVERTED;
-    }
-    format_gl = wined3d_format_gl(format);
-
-    wined3d_texture_gl_bind_and_dirtify(texture_gl, context_gl, srgb);
-
-    internal = wined3d_gl_get_internal_format(resource, format_gl, srgb);
-    if (!internal)
-        FIXME("No GL internal format for format %s.\n", debug_d3dformat(format->id));
-
-    TRACE("internal %#x, format %#x, type %#x.\n", internal, format_gl->format, format_gl->type);
-
-    if (wined3d_texture_use_immutable_storage(&texture_gl->t, gl_info))
-        wined3d_texture_gl_allocate_immutable_storage(texture_gl, internal, gl_info);
-    else
-        wined3d_texture_gl_allocate_mutable_storage(texture_gl, internal, format_gl, gl_info);
-    texture_gl->t.flags |= alloc_flag;
-}
-
-static void wined3d_texture_gl_prepare_rb(struct wined3d_texture_gl *texture_gl,
-        const struct wined3d_gl_info *gl_info, bool multisample)
-{
-    const struct wined3d_format_gl *format_gl;
-
-    format_gl = wined3d_format_gl(texture_gl->t.resource.format);
-    if (multisample)
-    {
-        if (texture_gl->rb_multisample)
-            return;
-
-        gl_info->fbo_ops.glGenRenderbuffers(1, &texture_gl->rb_multisample);
-        gl_info->fbo_ops.glBindRenderbuffer(GL_RENDERBUFFER, texture_gl->rb_multisample);
-        gl_info->fbo_ops.glRenderbufferStorageMultisample(GL_RENDERBUFFER,
-                wined3d_resource_get_sample_count(&texture_gl->t.resource),
-                format_gl->internal, texture_gl->t.resource.width, texture_gl->t.resource.height);
-        checkGLcall("glRenderbufferStorageMultisample()");
-        TRACE("Created multisample rb %u.\n", texture_gl->rb_multisample);
-    }
-    else
-    {
-        if (texture_gl->rb_resolved)
-            return;
-
-        gl_info->fbo_ops.glGenRenderbuffers(1, &texture_gl->rb_resolved);
-        gl_info->fbo_ops.glBindRenderbuffer(GL_RENDERBUFFER, texture_gl->rb_resolved);
-        gl_info->fbo_ops.glRenderbufferStorage(GL_RENDERBUFFER, format_gl->internal,
-                texture_gl->t.resource.width, texture_gl->t.resource.height);
-        checkGLcall("glRenderbufferStorage()");
-        TRACE("Created resolved rb %u.\n", texture_gl->rb_resolved);
-    }
-}
-
-static BOOL wined3d_texture_gl_prepare_location(struct wined3d_texture *texture,
-        unsigned int sub_resource_idx, struct wined3d_context *context, unsigned int location)
-{
-    struct wined3d_texture_gl *texture_gl = wined3d_texture_gl(texture);
-    struct wined3d_context_gl *context_gl = wined3d_context_gl(context);
-
-    switch (location)
-    {
-        case WINED3D_LOCATION_SYSMEM:
-            return texture->sub_resources[sub_resource_idx].user_memory ? TRUE
-                    : wined3d_resource_prepare_sysmem(&texture->resource);
-
-        case WINED3D_LOCATION_BUFFER:
-            wined3d_texture_gl_prepare_buffer_object(texture_gl, sub_resource_idx, context_gl);
-            return TRUE;
-
-        case WINED3D_LOCATION_TEXTURE_RGB:
-            wined3d_texture_gl_prepare_texture(texture_gl, context_gl, false);
-            return TRUE;
-
-        case WINED3D_LOCATION_TEXTURE_SRGB:
-            wined3d_texture_gl_prepare_texture(texture_gl, context_gl, true);
-            return TRUE;
-
-        case WINED3D_LOCATION_DRAWABLE:
-            if (!texture->swapchain)
-                ERR("Texture %p does not have a drawable.\n", texture);
-            return TRUE;
-
-        case WINED3D_LOCATION_RB_MULTISAMPLE:
-            wined3d_texture_gl_prepare_rb(texture_gl, context_gl->gl_info, true);
-            return TRUE;
-
-        case WINED3D_LOCATION_RB_RESOLVED:
-            wined3d_texture_gl_prepare_rb(texture_gl, context_gl->gl_info, false);
-            return TRUE;
-
-        default:
-            ERR("Invalid location %s.\n", wined3d_debug_location(location));
-            return FALSE;
-    }
-}
-
-static bool use_ffp_clear(const struct wined3d_texture *texture, unsigned int location)
-{
-    if (location == WINED3D_LOCATION_DRAWABLE)
-        return true;
-
-    if (location == WINED3D_LOCATION_TEXTURE_RGB
-            && !(texture->resource.format_caps & WINED3D_FORMAT_CAP_FBO_ATTACHABLE))
-        return false;
-    if (location == WINED3D_LOCATION_TEXTURE_SRGB
-            && !(texture->resource.format_caps & WINED3D_FORMAT_CAP_FBO_ATTACHABLE_SRGB))
-        return false;
-
-    return location & (WINED3D_LOCATION_RB_MULTISAMPLE | WINED3D_LOCATION_RB_RESOLVED
-            | WINED3D_LOCATION_TEXTURE_RGB | WINED3D_LOCATION_TEXTURE_SRGB);
-}
-
-static bool wined3d_texture_gl_clear(struct wined3d_texture *texture,
-        unsigned int sub_resource_idx, struct wined3d_context_gl *context_gl, unsigned int location)
-{
-    struct wined3d_texture_sub_resource *sub_resource = &texture->sub_resources[sub_resource_idx];
-    const struct wined3d_format *format = texture->resource.format;
-    const struct wined3d_gl_info *gl_info = context_gl->gl_info;
-    struct wined3d_bo_address addr;
-
-    /* The code that delays clears is Vulkan-specific, so here we should only
-     * encounter WINED3D_LOCATION_CLEARED on newly created resources and thus
-     * a zero clear value. */
-    if (!format->depth_size && !format->stencil_size)
-    {
-        if (sub_resource->clear_value.colour.r || sub_resource->clear_value.colour.g
-                || sub_resource->clear_value.colour.b || sub_resource->clear_value.colour.a)
-        {
-            ERR("Unexpected color clear value r=%08e, g=%08e, b=%08e, a=%08e.\n",
-                    sub_resource->clear_value.colour.r, sub_resource->clear_value.colour.g,
-                    sub_resource->clear_value.colour.b, sub_resource->clear_value.colour.a);
-        }
-    }
-    else
-    {
-        if (format->depth_size && sub_resource->clear_value.depth)
-            ERR("Unexpected depth clear value %08e.\n", sub_resource->clear_value.depth);
-        if (format->stencil_size && sub_resource->clear_value.stencil)
-            ERR("Unexpected stencil clear value %x.\n", sub_resource->clear_value.stencil);
-    }
-
-    if (use_ffp_clear(texture, location))
-    {
-        GLbitfield clear_mask = 0;
-
-        context_gl_apply_texture_draw_state(context_gl, texture, sub_resource_idx, location);
-
-        gl_info->gl_ops.gl.p_glDisable(GL_SCISSOR_TEST);
-        context_invalidate_state(&context_gl->c, STATE_RASTERIZER);
-
-        if (format->depth_size)
-        {
-            gl_info->gl_ops.gl.p_glDepthMask(GL_TRUE);
-            context_invalidate_state(&context_gl->c, STATE_DEPTH_STENCIL);
-
-            if (gl_info->gl_ops.ext.p_glClearDepthf)
-            {
-                GL_EXTCALL(glClearDepthf(0.0f));
-                clear_mask |= GL_DEPTH_BUFFER_BIT;
-            }
-            else if (!gl_info->is_gles)
-            {
-                gl_info->gl_ops.gl.p_glClearDepth(0.0);
-                clear_mask |= GL_DEPTH_BUFFER_BIT;
-            }
-            else if (gl_info->gl_ops.ext.p_glClearBufferfv)
-            {
-                static const GLfloat zero = 0.0f;
-
-                GL_EXTCALL(glClearBufferfv(GL_DEPTH, 0, &zero));
-            }
-            else
-            {
-                mgs_gles_clear_depth(context_gl, 0.0f);
-            }
-        }
-
-        if (format->stencil_size)
-        {
-            if (gl_info->supported[EXT_STENCIL_TWO_SIDE])
-                gl_info->gl_ops.gl.p_glDisable(GL_STENCIL_TEST_TWO_SIDE_EXT);
-            gl_info->gl_ops.gl.p_glStencilMask(~0u);
-            context_invalidate_state(&context_gl->c, STATE_DEPTH_STENCIL);
-            gl_info->gl_ops.gl.p_glClearStencil(0);
-            clear_mask |= GL_STENCIL_BUFFER_BIT;
-        }
-
-        if (!format->depth_size && !format->stencil_size)
-        {
-            gl_info->gl_ops.gl.p_glColorMask(GL_TRUE, GL_TRUE, GL_TRUE, GL_TRUE);
-            context_invalidate_state(&context_gl->c, STATE_BLEND);
-            gl_info->gl_ops.gl.p_glClearColor(0.0f, 0.0f, 0.0f, 0.0f);
-            clear_mask |= GL_COLOR_BUFFER_BIT;
-        }
-
-        gl_info->gl_ops.gl.p_glClear(clear_mask);
-        checkGLcall("clear texture");
-
-        wined3d_texture_validate_location(texture, sub_resource_idx, location);
-        return true;
-    }
-
-    if (!wined3d_texture_prepare_location(texture, sub_resource_idx, &context_gl->c, WINED3D_LOCATION_SYSMEM))
-        return false;
-    wined3d_texture_get_bo_address(texture, sub_resource_idx, &addr, WINED3D_LOCATION_SYSMEM);
-    memset(addr.addr, 0, sub_resource->size);
-    wined3d_texture_validate_location(texture, sub_resource_idx, WINED3D_LOCATION_SYSMEM);
-    return true;
-}
-
-/* Context activation is done by the caller. */
-static BOOL wined3d_texture_gl_load_location(struct wined3d_texture *texture,
-        unsigned int sub_resource_idx, struct wined3d_context *context, uint32_t location)
-{
-    struct wined3d_texture_sub_resource *sub_resource = &texture->sub_resources[sub_resource_idx];
-    struct wined3d_texture_gl *texture_gl = wined3d_texture_gl(texture);
-    struct wined3d_context_gl *context_gl = wined3d_context_gl(context);
-
-    TRACE("texture %p, sub_resource_idx %u, context %p, location %s.\n",
-            texture, sub_resource_idx, context, wined3d_debug_location(location));
-
-    if (!wined3d_texture_gl_prepare_location(texture, sub_resource_idx, context, location))
-        return FALSE;
-
-    if (sub_resource->locations & WINED3D_LOCATION_CLEARED)
-    {
-        if (!wined3d_texture_gl_clear(texture, sub_resource_idx, context_gl, location))
-            return FALSE;
-
-        if (sub_resource->locations & location)
-            return TRUE;
-    }
-
-    switch (location)
-    {
-        case WINED3D_LOCATION_SYSMEM:
-        case WINED3D_LOCATION_BUFFER:
-            return wined3d_texture_gl_load_sysmem(texture_gl, sub_resource_idx, context_gl, location);
-
-        case WINED3D_LOCATION_DRAWABLE:
-            return wined3d_texture_load_drawable(texture, sub_resource_idx, context);
-
-        case WINED3D_LOCATION_RB_RESOLVED:
-        case WINED3D_LOCATION_RB_MULTISAMPLE:
-            return wined3d_texture_load_renderbuffer(texture, sub_resource_idx, context, location);
-
-        case WINED3D_LOCATION_TEXTURE_RGB:
-        case WINED3D_LOCATION_TEXTURE_SRGB:
-            return wined3d_texture_gl_load_texture(texture_gl, sub_resource_idx,
-                    context_gl, location == WINED3D_LOCATION_TEXTURE_SRGB);
-
-        default:
-            FIXME("Unhandled %s load from %s.\n", wined3d_debug_location(location),
-                    wined3d_debug_location(texture->sub_resources[sub_resource_idx].locations));
-            return FALSE;
-    }
-}
-
-static void wined3d_texture_remove_buffer_object(struct wined3d_texture *texture,
-        unsigned int sub_resource_idx, struct wined3d_context_gl *context_gl)
-{
-    struct wined3d_texture_sub_resource *sub_resource = &texture->sub_resources[sub_resource_idx];
-    struct wined3d_bo_gl *bo_gl = wined3d_bo_gl(sub_resource->bo);
-
-    TRACE("texture %p, sub_resource_idx %u, context_gl %p.\n", texture, sub_resource_idx, context_gl);
-
-    wined3d_context_gl_destroy_bo(context_gl, bo_gl);
-    wined3d_texture_invalidate_location(texture, sub_resource_idx, WINED3D_LOCATION_BUFFER);
-    sub_resource->bo = NULL;
-    free(bo_gl);
-}
-
-static void wined3d_texture_gl_unload_location(struct wined3d_texture *texture,
-        struct wined3d_context *context, unsigned int location)
-{
-    struct wined3d_texture_gl *texture_gl = wined3d_texture_gl(texture);
-    struct wined3d_context_gl *context_gl = wined3d_context_gl(context);
-    struct wined3d_renderbuffer_entry *entry, *entry2;
-    unsigned int sub_count;
-
-    TRACE("texture %p, context %p, location %s.\n", texture, context, wined3d_debug_location(location));
-
-    switch (location)
-    {
-        case WINED3D_LOCATION_BUFFER:
-            sub_count = texture->level_count * texture->layer_count;
-            for (unsigned int i = 0; i < sub_count; ++i)
-            {
-                if (texture_gl->t.sub_resources[i].bo)
-                    wined3d_texture_remove_buffer_object(&texture_gl->t, i, context_gl);
-            }
-            break;
-
-        case WINED3D_LOCATION_TEXTURE_RGB:
-            if (texture_gl->texture_rgb.name)
-                gltexture_delete(texture_gl->t.resource.device, context_gl->gl_info, &texture_gl->texture_rgb);
-            break;
-
-        case WINED3D_LOCATION_TEXTURE_SRGB:
-            if (texture_gl->texture_srgb.name)
-                gltexture_delete(texture_gl->t.resource.device, context_gl->gl_info, &texture_gl->texture_srgb);
-            break;
-
-        case WINED3D_LOCATION_RB_MULTISAMPLE:
-            if (texture_gl->rb_multisample)
-            {
-                TRACE("Deleting multisample renderbuffer %u.\n", texture_gl->rb_multisample);
-                context_gl_resource_released(texture_gl->t.resource.device, texture_gl->rb_multisample, TRUE);
-                context_gl->gl_info->fbo_ops.glDeleteRenderbuffers(1, &texture_gl->rb_multisample);
-                texture_gl->rb_multisample = 0;
-            }
-            break;
-
-        case WINED3D_LOCATION_RB_RESOLVED:
-            LIST_FOR_EACH_ENTRY_SAFE(entry, entry2, &texture_gl->renderbuffers,
-                    struct wined3d_renderbuffer_entry, entry)
-            {
-                context_gl_resource_released(texture_gl->t.resource.device, entry->id, TRUE);
-                context_gl->gl_info->fbo_ops.glDeleteRenderbuffers(1, &entry->id);
-                list_remove(&entry->entry);
-                free(entry);
-            }
-            list_init(&texture_gl->renderbuffers);
-            texture_gl->current_renderbuffer = NULL;
-
-            if (texture_gl->rb_resolved)
-            {
-                TRACE("Deleting resolved renderbuffer %u.\n", texture_gl->rb_resolved);
-                context_gl_resource_released(texture_gl->t.resource.device, texture_gl->rb_resolved, TRUE);
-                context_gl->gl_info->fbo_ops.glDeleteRenderbuffers(1, &texture_gl->rb_resolved);
-                texture_gl->rb_resolved = 0;
-            }
-            break;
-
-        default:
-            ERR("Unhandled location %s.\n", wined3d_debug_location(location));
-            break;
-    }
-}
-
-static const struct wined3d_texture_ops texture_gl_ops =
-{
-    wined3d_texture_gl_prepare_location,
-    wined3d_texture_gl_load_location,
-    wined3d_texture_gl_unload_location,
-    wined3d_texture_gl_upload_data,
-    wined3d_texture_gl_download_data,
-};
-
-HRESULT wined3d_texture_gl_init(struct wined3d_texture_gl *texture_gl, struct wined3d_device *device,
-        const struct wined3d_resource_desc *desc, unsigned int layer_count, unsigned int level_count,
-        uint32_t flags, void *parent, const struct wined3d_parent_ops *parent_ops)
-{
-    const struct wined3d_gl_info *gl_info = &wined3d_adapter_gl(device->adapter)->gl_info;
-    HRESULT hr;
-
-    TRACE("texture_gl %p, device %p, desc %p, layer_count %u, "
-            "level_count %u, flags %#x, parent %p, parent_ops %p.\n",
-            texture_gl, device, desc, layer_count,
-            level_count, flags, parent, parent_ops);
-
-    if (!(desc->usage & WINED3DUSAGE_LEGACY_CUBEMAP) && layer_count > 1
-            && !gl_info->supported[EXT_TEXTURE_ARRAY])
-    {
-        WARN("OpenGL implementation does not support array textures.\n");
-        return WINED3DERR_INVALIDCALL;
-    }
-
-    switch (desc->resource_type)
-    {
-        case WINED3D_RTYPE_TEXTURE_1D:
-            if (layer_count > 1)
-                texture_gl->target = GL_TEXTURE_1D_ARRAY;
-            else
-                texture_gl->target = GL_TEXTURE_1D;
-            break;
-
-        case WINED3D_RTYPE_TEXTURE_2D:
-            if (desc->usage & WINED3DUSAGE_LEGACY_CUBEMAP)
-            {
-                texture_gl->target = GL_TEXTURE_CUBE_MAP_ARB;
-            }
-            else if (desc->multisample_type && gl_info->supported[ARB_TEXTURE_MULTISAMPLE])
-            {
-                if (layer_count > 1)
-                    texture_gl->target = GL_TEXTURE_2D_MULTISAMPLE_ARRAY;
-                else
-                    texture_gl->target = GL_TEXTURE_2D_MULTISAMPLE;
-            }
-            else
-            {
-                if (layer_count > 1)
-                    texture_gl->target = GL_TEXTURE_2D_ARRAY;
-                else
-                    texture_gl->target = GL_TEXTURE_2D;
-            }
-            break;
-
-        case WINED3D_RTYPE_TEXTURE_3D:
-            if (!gl_info->supported[EXT_TEXTURE3D])
-            {
-                WARN("OpenGL implementation does not support 3D textures.\n");
-                return WINED3DERR_INVALIDCALL;
-            }
-            texture_gl->target = GL_TEXTURE_3D;
-            break;
-
-        default:
-            ERR("Invalid resource type %s.\n", debug_d3dresourcetype(desc->resource_type));
-            return WINED3DERR_INVALIDCALL;
-    }
-
-    list_init(&texture_gl->renderbuffers);
-
-    if (FAILED(hr = wined3d_texture_init(&texture_gl->t, desc, layer_count, level_count,
-            flags, device, parent, parent_ops, &texture_gl[1], &texture_gl_ops)))
-        return hr;
-
-    if (texture_gl->target == GL_TEXTURE_2D_MULTISAMPLE_ARRAY || texture_gl->target == GL_TEXTURE_2D_MULTISAMPLE)
-        texture_gl->t.flags &= ~WINED3D_TEXTURE_DOWNLOADABLE;
-
-    return WINED3D_OK;
-}
-
-void wined3d_gl_texture_swizzle_from_color_fixup(GLint swizzle[4], struct color_fixup_desc fixup)
+static void wined3d_texture_gl_download_data(struct wined3d_context *context,
+        struct wined3d_texture *src_texture, unsigned int src_sub_resource_idx, unsigned int src_location,
+        const struct wined3d_box *src_box, const struct wined3d_bo_address *dst_bo_addr,
+        const struct wined3d_format *dst_format, unsigned int dst_x, unsigned int dst_y, unsigned int dst_z,
+        unsigned int dst_row_pitch, unsigned int dst_slice_pitch)
 {
-    static const GLenum swizzle_source[] =
-    {
-        GL_ZERO,  /* CHANNEL_SOURCE_ZERO */
-        GL_ONE,   /* CHANNEL_SOURCE_ONE */
-        GL_RED,   /* CHANNEL_SOURCE_X */
-        GL_GREEN, /* CHANNEL_SOURCE_Y */
-        GL_BLUE,  /* CHANNEL_SOURCE_Z */
-        GL_ALPHA, /* CHANNEL_SOURCE_W */
-    };
-
-    swizzle[0] = swizzle_source[fixup.x_source];
-    swizzle[1] = swizzle_source[fixup.y_source];
-    swizzle[2] = swizzle_source[fixup.z_source];
-    swizzle[3] = swizzle_source[fixup.w_source];
+    wined3d_texture_gl_download_data_internal(context, src_texture, src_sub_resource_idx, src_location,
+            src_box, dst_bo_addr, dst_format, dst_x, dst_y, dst_z,
+            dst_row_pitch, dst_slice_pitch);
 }
 
 /* Context activation is done by the caller. */
-void wined3d_texture_gl_bind(struct wined3d_texture_gl *texture_gl,
-        struct wined3d_context_gl *context_gl, BOOL srgb)
-{
-    const struct wined3d_format *format = texture_gl->t.resource.format;
-    const struct wined3d_gl_info *gl_info = context_gl->gl_info;
-    const struct color_fixup_desc fixup = format->color_fixup;
-    struct gl_texture *gl_tex;
-    GLenum target;
-
-    TRACE("texture_gl %p, context_gl %p, srgb %#x.\n", texture_gl, context_gl, srgb);
-
-    if (!needs_separate_srgb_gl_texture(&context_gl->c, &texture_gl->t))
-        srgb = FALSE;
-
-    /* sRGB mode cache for preload() calls outside drawprim. */
-    if (srgb)
-        texture_gl->t.flags |= WINED3D_TEXTURE_IS_SRGB;
-    else
-        texture_gl->t.flags &= ~WINED3D_TEXTURE_IS_SRGB;
-
-    gl_tex = wined3d_texture_gl_get_gl_texture(texture_gl, srgb);
-    target = texture_gl->target;
-
-    if (gl_tex->name)
-    {
-        wined3d_context_gl_bind_texture(context_gl, target, gl_tex->name);
-        return;
-    }
-
-    gl_info->gl_ops.gl.p_glGenTextures(1, &gl_tex->name);
-    checkGLcall("glGenTextures");
-    TRACE("Generated texture %d.\n", gl_tex->name);
-
-    if (!gl_tex->name)
-    {
-        ERR("Failed to generate a texture name.\n");
-        return;
-    }
-
-    /* Initialise the state of the texture object to the OpenGL defaults, not
-     * the wined3d defaults. */
-    gl_tex->sampler_desc.address_u = WINED3D_TADDRESS_WRAP;
-    gl_tex->sampler_desc.address_v = WINED3D_TADDRESS_WRAP;
-    gl_tex->sampler_desc.address_w = WINED3D_TADDRESS_WRAP;
-    memset(gl_tex->sampler_desc.border_color, 0, sizeof(gl_tex->sampler_desc.border_color));
-    gl_tex->sampler_desc.mag_filter = WINED3D_TEXF_LINEAR;
-    gl_tex->sampler_desc.min_filter = WINED3D_TEXF_POINT; /* GL_NEAREST_MIPMAP_LINEAR */
-    gl_tex->sampler_desc.mip_filter = WINED3D_TEXF_LINEAR; /* GL_NEAREST_MIPMAP_LINEAR */
-    gl_tex->sampler_desc.lod_bias = 0.0f;
-    gl_tex->sampler_desc.min_lod = -1000.0f;
-    gl_tex->sampler_desc.max_lod = 1000.0f;
-    gl_tex->sampler_desc.max_anisotropy = 1;
-    gl_tex->sampler_desc.reduction_mode = WINED3D_FILTER_REDUCTION_WEIGHTED_AVERAGE;
-    gl_tex->sampler_desc.comparison_func = WINED3D_CMP_LESSEQUAL;
-    if (gl_info->supported[EXT_TEXTURE_SRGB_DECODE])
-        gl_tex->sampler_desc.srgb_decode = TRUE;
-    else
-        gl_tex->sampler_desc.srgb_decode = srgb;
-    gl_tex->sampler_desc.mip_base_level = 0;
-    wined3d_texture_set_dirty(&texture_gl->t);
-
-    wined3d_context_gl_bind_texture(context_gl, target, gl_tex->name);
-
-    /* For a new texture we have to set the texture levels after binding the
-     * texture. */
-    TRACE("Setting GL_TEXTURE_MAX_LEVEL to %u.\n", texture_gl->t.level_count - 1);
-    gl_info->gl_ops.gl.p_glTexParameteri(target, GL_TEXTURE_MAX_LEVEL, texture_gl->t.level_count - 1);
-    checkGLcall("glTexParameteri(target, GL_TEXTURE_MAX_LEVEL, texture->level_count)");
-
-    if (target == GL_TEXTURE_CUBE_MAP_ARB)
-    {
-        /* Cubemaps are always set to clamp, regardless of the sampler state. */
-        gl_info->gl_ops.gl.p_glTexParameteri(target, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
-        gl_info->gl_ops.gl.p_glTexParameteri(target, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
-        gl_info->gl_ops.gl.p_glTexParameteri(target, GL_TEXTURE_WRAP_R, GL_CLAMP_TO_EDGE);
-    }
-
-    if (texture_gl->t.flags & WINED3D_TEXTURE_COND_NP2 && target != GL_TEXTURE_2D_MULTISAMPLE
-            && target != GL_TEXTURE_2D_MULTISAMPLE_ARRAY)
-    {
-        /* Conditional non power of two textures use a different clamping
-         * default. If we're using the GL_WINE_normalized_texrect partial
-         * driver emulation, we're dealing with a GL_TEXTURE_2D texture which
-         * has the address mode set to repeat - something that prevents us
-         * from hitting the accelerated codepath. Thus manually set the GL
-         * state. The same applies to filtering. Even if the texture has only
-         * one mip level, the default LINEAR_MIPMAP_LINEAR filter causes a SW
-         * fallback on macos. */
-        gl_info->gl_ops.gl.p_glTexParameteri(target, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
-        gl_info->gl_ops.gl.p_glTexParameteri(target, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
-        gl_info->gl_ops.gl.p_glTexParameteri(target, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
-        gl_info->gl_ops.gl.p_glTexParameteri(target, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
-        checkGLcall("glTexParameteri");
-        gl_tex->sampler_desc.address_u = WINED3D_TADDRESS_CLAMP;
-        gl_tex->sampler_desc.address_v = WINED3D_TADDRESS_CLAMP;
-        gl_tex->sampler_desc.mag_filter = WINED3D_TEXF_POINT;
-        gl_tex->sampler_desc.min_filter = WINED3D_TEXF_POINT;
-        gl_tex->sampler_desc.mip_filter = WINED3D_TEXF_NONE;
-    }
-
-    if (gl_info->supported[WINED3D_GL_LEGACY_CONTEXT] && gl_info->supported[ARB_DEPTH_TEXTURE])
-    {
-        gl_info->gl_ops.gl.p_glTexParameteri(target, GL_DEPTH_TEXTURE_MODE_ARB, GL_INTENSITY);
-        checkGLcall("glTexParameteri(GL_DEPTH_TEXTURE_MODE_ARB, GL_INTENSITY)");
-    }
-
-    if (!is_identity_fixup(fixup) && can_use_texture_swizzle(context_gl->c.d3d_info, format))
-    {
-        GLint swizzle[4];
-
-        wined3d_gl_texture_swizzle_from_color_fixup(swizzle, fixup);
-        gl_info->gl_ops.gl.p_glTexParameteriv(target, GL_TEXTURE_SWIZZLE_RGBA, swizzle);
-        checkGLcall("set format swizzle");
-    }
-}
-
-/* Context activation is done by the caller. */
-void wined3d_texture_gl_bind_and_dirtify(struct wined3d_texture_gl *texture_gl,
-        struct wined3d_context_gl *context_gl, BOOL srgb)
-{
-    /* FIXME: Ideally we'd only do this when touching a binding that's used by
-     * a shader. */
-    context_invalidate_compute_state(&context_gl->c, STATE_COMPUTE_SHADER_RESOURCE_BINDING);
-    context_invalidate_state(&context_gl->c, STATE_GRAPHICS_SHADER_RESOURCE_BINDING);
-
-    wined3d_texture_gl_bind(texture_gl, context_gl, srgb);
-}
-
-/* Context activation is done by the caller (state handler). */
-/* This function relies on the correct texture being bound and loaded. */
-void wined3d_texture_gl_apply_sampler_desc(struct wined3d_texture_gl *texture_gl,
-        const struct wined3d_sampler_desc *sampler_desc, const struct wined3d_context_gl *context_gl)
-{
-    const struct wined3d_gl_info *gl_info = context_gl->gl_info;
-    GLenum target = texture_gl->target;
-    struct gl_texture *gl_tex;
-    unsigned int state;
-
-    TRACE("texture_gl %p, sampler_desc %p, context_gl %p.\n", texture_gl, sampler_desc, context_gl);
-
-    gl_tex = wined3d_texture_gl_get_gl_texture(texture_gl, texture_gl->t.flags & WINED3D_TEXTURE_IS_SRGB);
-
-    state = sampler_desc->address_u;
-    if (state != gl_tex->sampler_desc.address_u)
-    {
-        gl_info->gl_ops.gl.p_glTexParameteri(target, GL_TEXTURE_WRAP_S,
-                gl_info->wrap_lookup[state - WINED3D_TADDRESS_WRAP]);
-        gl_tex->sampler_desc.address_u = state;
-    }
-
-    state = sampler_desc->address_v;
-    if (state != gl_tex->sampler_desc.address_v)
-    {
-        gl_info->gl_ops.gl.p_glTexParameteri(target, GL_TEXTURE_WRAP_T,
-                gl_info->wrap_lookup[state - WINED3D_TADDRESS_WRAP]);
-        gl_tex->sampler_desc.address_v = state;
-    }
-
-    state = sampler_desc->address_w;
-    if (state != gl_tex->sampler_desc.address_w)
-    {
-        gl_info->gl_ops.gl.p_glTexParameteri(target, GL_TEXTURE_WRAP_R,
-                gl_info->wrap_lookup[state - WINED3D_TADDRESS_WRAP]);
-        gl_tex->sampler_desc.address_w = state;
-    }
-
-    if (memcmp(gl_tex->sampler_desc.border_color, sampler_desc->border_color,
-            sizeof(gl_tex->sampler_desc.border_color)))
-    {
-        gl_info->gl_ops.gl.p_glTexParameterfv(target, GL_TEXTURE_BORDER_COLOR, &sampler_desc->border_color[0]);
-        memcpy(gl_tex->sampler_desc.border_color, sampler_desc->border_color,
-                sizeof(gl_tex->sampler_desc.border_color));
-    }
-
-    state = sampler_desc->mag_filter;
-    if (state != gl_tex->sampler_desc.mag_filter)
-    {
-        gl_info->gl_ops.gl.p_glTexParameteri(target, GL_TEXTURE_MAG_FILTER, wined3d_gl_mag_filter(state));
-        gl_tex->sampler_desc.mag_filter = state;
-    }
-
-    if (sampler_desc->min_filter != gl_tex->sampler_desc.min_filter
-            || sampler_desc->mip_filter != gl_tex->sampler_desc.mip_filter)
-    {
-        gl_info->gl_ops.gl.p_glTexParameteri(target, GL_TEXTURE_MIN_FILTER,
-                wined3d_gl_min_mip_filter(sampler_desc->min_filter, sampler_desc->mip_filter));
-        gl_tex->sampler_desc.min_filter = sampler_desc->min_filter;
-        gl_tex->sampler_desc.mip_filter = sampler_desc->mip_filter;
-    }
-
-    state = sampler_desc->max_anisotropy;
-    if (state != gl_tex->sampler_desc.max_anisotropy)
-    {
-        if (gl_info->supported[ARB_TEXTURE_FILTER_ANISOTROPIC])
-            gl_info->gl_ops.gl.p_glTexParameteri(target, GL_TEXTURE_MAX_ANISOTROPY, state);
-        else
-            WARN("Anisotropic filtering not supported.\n");
-        gl_tex->sampler_desc.max_anisotropy = state;
-    }
-
-    if (!sampler_desc->srgb_decode != !gl_tex->sampler_desc.srgb_decode
-            && (context_gl->c.d3d_info->wined3d_creation_flags & WINED3D_SRGB_READ_WRITE_CONTROL)
-            && gl_info->supported[EXT_TEXTURE_SRGB_DECODE])
-    {
-        gl_info->gl_ops.gl.p_glTexParameteri(target, GL_TEXTURE_SRGB_DECODE_EXT,
-                sampler_desc->srgb_decode ? GL_DECODE_EXT : GL_SKIP_DECODE_EXT);
-        gl_tex->sampler_desc.srgb_decode = sampler_desc->srgb_decode;
-    }
-
-    if (sampler_desc->reduction_mode != gl_tex->sampler_desc.reduction_mode)
-    {
-        GLenum reduction_mode = wined3d_gl_filter_reduction_mode(sampler_desc->reduction_mode);
-
-        if (sampler_desc->reduction_mode == WINED3D_FILTER_REDUCTION_COMPARISON)
-            gl_info->gl_ops.gl.p_glTexParameteri(target, GL_TEXTURE_COMPARE_MODE_ARB, GL_COMPARE_R_TO_TEXTURE_ARB);
-        else
-            gl_info->gl_ops.gl.p_glTexParameteri(target, GL_TEXTURE_COMPARE_MODE_ARB, GL_NONE);
-
-        if (gl_info->supported[ARB_TEXTURE_FILTER_MINMAX])
-            gl_info->gl_ops.gl.p_glTexParameteri(target, GL_TEXTURE_REDUCTION_MODE_ARB, reduction_mode);
-        else if (reduction_mode != GL_WEIGHTED_AVERAGE_ARB)
-            WARN("Sampler min/max reduction filtering is not supported.\n");
-
-        gl_tex->sampler_desc.reduction_mode = sampler_desc->reduction_mode;
-    }
-
-    checkGLcall("Texture parameter application");
-
-    if (gl_info->supported[EXT_TEXTURE_LOD_BIAS])
-    {
-        gl_info->gl_ops.gl.p_glTexEnvf(GL_TEXTURE_FILTER_CONTROL_EXT,
-                GL_TEXTURE_LOD_BIAS_EXT, sampler_desc->lod_bias);
-        checkGLcall("glTexEnvf(GL_TEXTURE_LOD_BIAS_EXT, ...)");
-    }
-}
-
-/* In D3D the depth stencil dimensions have to be greater than or equal to the
- * render target dimensions. With FBOs, the dimensions have to be an exact match. */
-/* TODO: We should synchronize the renderbuffer's content with the texture's content. */
 /* Context activation is done by the caller. */
-void wined3d_texture_gl_set_compatible_renderbuffer(struct wined3d_texture_gl *texture_gl,
-        struct wined3d_context_gl *context_gl, unsigned int level, const struct wined3d_rendertarget_info *rt)
-{
-    const struct wined3d_gl_info *gl_info = context_gl->gl_info;
-    struct wined3d_renderbuffer_entry *entry;
-    unsigned int src_width, src_height;
-    unsigned int width, height;
-    GLuint renderbuffer = 0;
-
-    if (gl_info->supported[ARB_FRAMEBUFFER_OBJECT])
-        return;
-
-    if (rt && rt->resource->format->id != WINED3DFMT_NULL)
-    {
-        struct wined3d_texture *rt_texture;
-        unsigned int rt_level;
-
-        if (rt->resource->type == WINED3D_RTYPE_BUFFER)
-        {
-            FIXME("Unsupported resource type %s.\n", debug_d3dresourcetype(rt->resource->type));
-            return;
-        }
-        rt_texture = wined3d_texture_from_resource(rt->resource);
-        rt_level = rt->sub_resource_idx % rt_texture->level_count;
-
-        width = wined3d_texture_get_level_width(rt_texture, rt_level);
-        height = wined3d_texture_get_level_height(rt_texture, rt_level);
-    }
-    else
-    {
-        width = wined3d_texture_get_level_width(&texture_gl->t, level);
-        height = wined3d_texture_get_level_height(&texture_gl->t, level);
-    }
-
-    src_width = wined3d_texture_get_level_width(&texture_gl->t, level);
-    src_height = wined3d_texture_get_level_height(&texture_gl->t, level);
-
-    /* A depth stencil smaller than the render target is not valid */
-    if (width > src_width || height > src_height)
-        return;
-
-    /* Remove any renderbuffer set if the sizes match */
-    if (width == src_width && height == src_height)
-    {
-        texture_gl->current_renderbuffer = NULL;
-        return;
-    }
-
-    /* Look if we've already got a renderbuffer of the correct dimensions */
-    LIST_FOR_EACH_ENTRY(entry, &texture_gl->renderbuffers, struct wined3d_renderbuffer_entry, entry)
-    {
-        if (entry->width == width && entry->height == height)
-        {
-            renderbuffer = entry->id;
-            texture_gl->current_renderbuffer = entry;
-            break;
-        }
-    }
-
-    if (!renderbuffer)
-    {
-        const struct wined3d_format_gl *format_gl;
-
-        format_gl = wined3d_format_gl(texture_gl->t.resource.format);
-        gl_info->fbo_ops.glGenRenderbuffers(1, &renderbuffer);
-        gl_info->fbo_ops.glBindRenderbuffer(GL_RENDERBUFFER, renderbuffer);
-        gl_info->fbo_ops.glRenderbufferStorage(GL_RENDERBUFFER, format_gl->internal, width, height);
-
-        entry = malloc(sizeof(*entry));
-        entry->width = width;
-        entry->height = height;
-        entry->id = renderbuffer;
-        list_add_head(&texture_gl->renderbuffers, &entry->entry);
-
-        texture_gl->current_renderbuffer = entry;
-    }
-
-    checkGLcall("set compatible renderbuffer");
-}
